Added clipboard visualization (including transformations) + some code improvements

DustVoice opened the pull request:
Of course some clever guy will probably point out some errors. But as the plugin has added fully functional 

- Clipboard display
- Clipboard display with applied transform
- Schematics display
- show particles for all players. Can be disabled with ingame command `/wesv showForAllPlayers disable`

With the same code style used in the original plugin I see no issue with starting a pull request. I am more than happy to do some additional stuff.

I will provide some examples:

Currently only
- clipboardParticleEffect
- clipboardParticleData
- updateClipboardParticlesInterval
and
- showForAllPlayers

have been implemented as additional configuration options. There could be several more like custom gabs for the clipboard, custom maxSize, etc.

But I found these clipboard specific settings to be the most reasonable.

As I mentioned above I have also added a command to change the `showForAllPlayers` property, if it gets too messy. There could be several more such ingame commands like a custom particle for each player, custom gaps, etc. to adjust without reloading. For example you would need to have a larger gap if you select > 10000 blocks, but a smaller gap for 5 blocks. Hope this makes sense.

Of course you could then also have custom config files for each player and so on.

But this was neither the goal nor should it be.

One step after another. But I will definetely keep working on this plugin.

DustVoice commented:
I will remove the unused imports tomorrow. Please let me know if you have any other issue so I can work on fixing it

DustVoice commented:
I'll fix the codacy bit issues tomorrow and then update on my progress. I'm aiming to rewrite the command line handling as well

DustVoice commented:
I have to look into one issue that appeared

Somehow expanding a region doesn't update the particles anymore. I'm gonna look into this. Until I got this fixed, don't merge it just yet

DustVoice commented:
Ok so apparently it was just my server. But please as always test it please

DustVoice commented:
Apparently the problem only occurs when "checkForAxe" is enabled. Switching to another item and back to the axe displays the correct region

MrMicky-FR commented:
Thanks for your pull request !

Implementing clipboard visualization is a great idea, however the current implementation add a lot of duplicate code.

And did you test this with WorldEdit 6 on older Minecraft versions ?

DustVoice commented:
Ok so I added support for visualization of clipboard transformations for **all** selection types. Even for the ConvexPolyhedralSelection. It is all correctly displayed. I also fixed the issues of the Codacy bot except the "TooManyFields" error for WorldEditSelectionVisualizer, which isn't really neccesary. I'm gonna fix the last codacy bot errors. Tomorrow I will check again if the plugin runs smoothly with all kinds of minecraft/worldedit versions. But at the moment everything seems to be fine. I will probably now work on implementing more commands and restructuring the code. Some classes have to be duplicated by normal coding standards though. The hasChanged Event for example is only really of use if there are different Event Classes for each different event.

DustVoice commented on .gitignore:
```diff
@@ -17,4 +17,7 @@
+dependency-reduced-pom.xml
+
+# Own schematics loader (not used currently but can be useful later; ugly as hell)
+plugin/src/main/java/com/rojel/wesv/SchematicsHandler.txt
```
This schematics loader was completely written from scratch by me, as I thought the standard schematics loader was the issue that loaded schematics weren't displayed with particles. After 3h I saw my mistake. So this is just left over. Nevertheless I decided to not commit this ugly file.

DustVoice commented on plugin/src/main/java/com/rojel/wesv/ParticleTask.java:
```diff
@@ -1,45 +0,0 @@
-package com.rojel.wesv;
```
Moved to another package

DustVoice commented:
Also I saw you FastInv class in you repositories. If I understood correctly it enables custom Inventory displays with custom Items, etc... Maybe it would be nice to implement ingame config changes with this. If you have kinda like an chest inventory and by clicking the items you enable/disable values or you can then type in chat your new value, and so on. Just a random thought.

MrMicky-FR commented:
Thanks again for your pull request and sorry for the really long delays.

Implementing clipboard display is a great feature, but theses changes would have added lot of duplicated code, but I don't see how it could be possible to do it in another way.

This is why I think the best solution was to rewrite the plugin, to use more maintenable code and implements somes other features too that would be not possible to correctly add with the current code without lot of duplications (also some features like per-player settings with an inventory should be easier to add).

Currently only clipboard display is implemented from this pull request, in a slighty different way: the clipboard points are calculated in the update task but the particles are displayed relative to the player positions in the particle task, so the particles always spawn at the location even if the player move

Also there was some compatibility issues because all vectors in the `com.sk89q.worldedit.math` package was introduced in WorldEdit 7 wich only works on 1.13/1.14

For the show particles to all players in wich case it can be used ? Because this would spawn really big amount of particles if few players are using we a the same time

I also changed the default config to use the settings you recommended and it's works great 👍

Sorry for the time I take to respond and for not merging this. Even if I closed this pull request, it was helpfull 😃

DustVoice commented:
Thank you for your response. If I understood correctly, you basically modified my code with all the math to display transformations and so on and stuff to match your new plug in infrastructure. Is that correct? Also the thing about "all player visibility" was a feature we always wanted but never had: I can copy the selection to my clipboard and my girlfriend can give me directions where to paste it in order for it to align perfectly (as in go 2 blocks right and 2 forward) because I can't see it with big clipboard contents. We would love to update our server but in order for us to take this step the wesv plug in would need to have these features:
- base features of the plug in before I started the whole clipboard visualization stuff
- clipboard visualization
- visualization of clipboard transforms (mirror, rotate, etc of basic shapes like cubes cylinders spheres etc)
- OPTION to enable all players to view the particles of other players. I specifically mention the option to do that because as you have said, in this case you need to have the other people in a discord or smth like that to tell them not to use worldedit at the same time. 

If you could tell me if the current plug in version has all of this I would be more than happy to update everything immediately. 

Until then I will probably use my stone age version

DustVoice commented:
Also did you mean the particles stay at the position? This wouldn't make sense to me as the clipboard does move relative to the player...

MrMicky-FR commented:
For the base features of the plugin they should all be still available (only the FAWE fuzzy selection display was removed), the clipboard visualization works.
But the clipboard transforms and the option to show particles to all players is not supported yet.

The only small problem to add the clipboard transform is to support WorldEdit 6 and WorldEdit 7 it will require an important amount of compatiblity to works with the changes in vectors between the two api versions

> Also did you mean the particles stay at the position? This wouldn't make sense to me as the clipboard does move relative to the player...

It's the opposite: the particles are always at the good position, even if the clipboard is not updated
